intra/dialers/split_and_desync.go

```diff
@@ -0,0 +1,347 @@
+				}
+			}
+		} else {
+			if exceedTTL(cmsgArr) {
+				fromPort := from.(*unix.SockaddrInet4).Port
+				ttl = fromPort - basePort
+				if ttl > split1.TTL && ttl <= maxTTL {
+					split1.TTL = ttl
+				}
+			}
+		}
+	}
+
+	return split1, nil
+}
+
+func DialWithSplitAndDesyncFixedTtl(d *protect.RDial, addr *net.TCPAddr, initialTTL int, payload []byte) (DuplexConn, error) {
+	tcpConn, err := d.DialTCP(addr.Network(), nil, addr)
+	if err != nil {
+		return nil, err
+	}
+	if tcpConn == nil {
+		return nil, errNoConn
+	}
+	split1 := &OverwriteSplitter{
+		TCPConn: tcpConn,
+		TTL:     initialTTL,
+		Payload: payload,
+	}
+	return split1, nil
+}
+
+func DialWithSplitAndDesyncSmart(d *protect.RDial, addr *net.TCPAddr, maxTTL int, payload []byte) (DuplexConn, error) {
+	ttl, ok := queryTracerouteResult(addr.IP)
+	if ok {
+		return DialWithSplitAndDesyncFixedTtl(d, addr, ttl, payload)
+	}
+	conn, err := DialWithSplitAndDesyncTraceroute(d, addr, maxTTL, payload)
+	if err == nil {
+		addTracerouteResult(addr.IP, conn.(*OverwriteSplitter).TTL)
+	}
+	return conn, err
+}
+
+func (s *OverwriteSplitter) Write(b []byte) (int, error) {
+	conn := s.TCPConn
+	if s.Used {
+		// After the first write, there is no special write behavior.
+		return conn.Write(b)
+	}
+
+	// Setting `Used` to true ensures that this code only runs once per socket.
+	s.Used = true
+	if len(b) <= len(s.Payload) {
+		return conn.Write(b)
+	}
+	rawConn, err := conn.SyscallConn()
+	if err != nil {
+		return 0, err
+	}
+	if rawConn == nil {
+		return 0, errors.New("(*TCPConn) SyscallConn() returned nil")
+	}
+	var sockFD int
+	err = rawConn.Control(func(fd uintptr) {
+		sockFD = int(fd)
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	fileFD, err := unix.MemfdCreate("haar", unix.O_RDWR)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		err := unix.Close(fileFD)
+		if err != nil {
+			log.E("desync: cannot close memfd: %v", err)
+		}
+	}()
+	err = unix.Ftruncate(fileFD, int64(len(s.Payload)))
+	if err != nil {
+		return 0, err
+	}
+	firstSegment, err := unix.Mmap(fileFD, 0, len(s.Payload), unix.PROT_WRITE, unix.MAP_SHARED)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		err := unix.Munmap(firstSegment)
+		if err != nil {
+			log.E("desync: Munmap failed: %v", err)
+		}
+	}()
+
+	// We want s.Payload to be seen by censors, but don't want s.Payload to be seen by the server.
+	copy(firstSegment, s.Payload)
+	mRemote := conn.RemoteAddr()
+	if mRemote == nil {
+		return 0, errors.New("RemoteAddr() returned nil")
+	}
+	isIPv6 := strings.Contains(mRemote.String(), "[")
+	if isIPv6 {
+		err = unix.SetsockoptInt(sockFD, unix.IPPROTO_IPV6, unix.IPV6_UNICAST_HOPS, s.TTL)
+	} else {
+		err = unix.SetsockoptInt(sockFD, unix.IPPROTO_IP, unix.IP_TTL, s.TTL)
+	}
+	if err != nil {
+		return 0, err
+	}
+	var offset int64 = 0
+	n1, err := unix.Sendfile(sockFD, fileFD, &offset, len(s.Payload))
+	if err != nil {
+		return n1, err
+	}
+
+	copy(firstSegment, b[:len(s.Payload)])
```

Comment:
@Lanius-collaris what writes this newly copied to `firstSegment`? Or, is it not required and written to in a retransmission (somehow...)?